lib/groupmem.c

```diff
@@ -77,7 +77,8 @@ void gr_free_members (struct group *grent)
 	}
 }
 
-void gr_free (/*@out@*/ /*@only@*/struct group *grent)
+void
+gr_free(/*@only@*/struct group *grent)
 {
```

Comment:
Hm.  I don't see what your commit message has to do with removing the splint /*@only@*/ annotation.  At the same time, I don't see that the current annotations are correct - https://splint.org/manual/manual.html#operation says

"Splint does not report an error when a pointer to allocated but undefined storage is passed as an out parameter.  Within the body of a function, Splint will assume an out parameter is allocated but not necessarily bound to a value, so an error is reported if its value is used before it is defined."

...  So dropping these is probably the right thing, but I'm still thinking about it.

Reply:
> Hm. I don't see what your commit message has to do with removing the splint /_@only@_/ annotation.

You probably misread, or I'm misunderstanding.  It's about removing the `/@out@/` comment, not the `/@only@/` one.

AFAICS, I'm not removing `/@only@/` anywhere in the patch set:

```diff
$ git diff --word-diff-regex=. master..access --color=always | grep only
@@ -36,7 +36,8 @@ static /*@null@*/ /*@only@*/void *group_dup (const void *ent)
/*@only@*/[- /*@out@*/-]char **
add_list[- -](/*@returned@*/ /*@only@*/char **list, const char *member)
/*@only@*/[- /*@out@*/-]char **
del_list[- -](/*@returned@*/ /*@only@*/char **list, const char *member)
/*@only@*/[- /*@out@*/-]char **
extern /*@only@*//*@null@*/struct group *getgr_nam_gid (/*@null@*/const char *grname);
extern /*@null@*/ /*@only@*/struct group *__gr_dup (const struct group *grent);
extern /*@only@*/[- /*@out@*/-]char **add_list (/*@returned@*/ /*@only@*/char **, const char *);
extern /*@only@*/[- /*@out@*/-]char **del_list (/*@returned@*/ /*@only@*/char **, const char *);
extern /*@only@*/[- /*@out@*/-]char **dup_list (char *const *);
extern /*@only@*/char **comma_to_list (const char *);
extern /*@null@*/ /*@only@*/struct passwd *__pw_dup (const struct passwd *pwent);
extern /*@null@*/ /*@only@*/struct sgrp *__sgr_dup (const struct sgrp *sgent);
extern /*@null@*/ /*@only@*/struct spwd *__spw_dup (const struct spwd *spent);
@@ -26,7 +26,8 @@ static /*@null@*/ /*@only@*/void *passwd_dup (const void *ent)
@@ -117,14 +117,16 @@ static /*@null@*/ /*@only@*/void *gshadow_dup (const void *ent)
@@ -31,7 +31,8 @@ static /*@null@*/ /*@only@*/void *shadow_dup (const void *ent)
@@ -56,7 +56,8 @@ static /*@null@*/ /*@only@*/void *subordinate_dup (const void *ent)
```

And a sanity check that I didn't mistype in the commit log and say `/@only@/` when I meant `/@out@/`:

```sh
$ git log master..access | grep '\<only\>'
$
```

I'll assume it's a typo and you meant `/@out@/`, I'll rewrite the quote:

```
Hm. I don't see what your commit message has to do with removing the splint /@out@/ (typo fixed) annotation.
```



Let's copy the commit here (limited to this file, for simplification, and with extra context lines, for showing the point):

```diff
$ git show -U10 f213c354 -- lib/groupmem.c
commit f213c35431a15f49df7b2922140c70ad58e020dd
Author: Alejandro Colomar <alx@kernel.org>
Date:   Mon Dec 11 14:01:38 2023 +0100

    lib/: Remove /*@out@*/ comment from functions that read the pointee
    
    These functions (e.g., gr_free()), explicitly dereference the pointer
    and read the pointee.  The [[gnu::access(write_only, ...)]] attribute
    means that the pointee can be uninitialized, since it won't read it.
    That's appropriate for free(3), which does not read the memory it frees
    (however, free(3) does not read _nor_ write to the memory, so it would
    be inappropriate to use write_only with it).  However, our functions
    don't behave like free(3), and they do read the memory.
    
    Signed-off-by: Alejandro Colomar <alx@kernel.org>

diff --git a/lib/groupmem.c b/lib/groupmem.c
index eb3348aa..69d4435b 100644
--- a/lib/groupmem.c
+++ b/lib/groupmem.c
@@ -70,20 +70,21 @@ void gr_free_members (struct group *grent)
        if (NULL != grent->gr_mem) {
                size_t i;
                for (i = 0; NULL != grent->gr_mem[i]; i++) {
                        free (grent->gr_mem[i]);
                }
                free (grent->gr_mem);
                grent->gr_mem = NULL;
        }
 }
 
-void gr_free (/*@out@*/ /*@only@*/struct group *grent)
+void
+gr_free(/*@only@*/struct group *grent)
 {
        free (grent->gr_name);
        if (NULL != grent->gr_passwd) {
                strzero (grent->gr_passwd);
                free (grent->gr_passwd);
        }
        gr_free_members(grent);
        free (grent);
 }
```

As you can see above, gr_free() reads `*grent`.  To be precise, it reads `(*grent).gr_name` (a.k.a. `grent->gr_name`).  This means that if it is passed a pointer to unallocated memory, this function will read garbage, invoking nasal demons^W^Wundefined behavior.  Just imagine what would happen if you call the following code:

```c
gr_free(malloc(sizeof(struct group)));
```

> At the same time, I don't see that the current annotations are correct - https://splint.org/manual/manual.html#operation says

Aha!  I didn't know what program used these annotations; I was firing blindly.  Thanks for the pointer!

> 
> "Splint does not report an error when a pointer to allocated but undefined storage is passed as an out parameter. Within the body of a function, Splint will assume an out parameter is allocated but not necessarily bound to a value, so an error is reported if its value is used before it is defined."
> 
> ... So dropping these is probably the right thing, but I'm still thinking about it.

Maybe the commit message sounds better in my head because I wrote it and tend to interpret it the way I want it to be read.  If you think it doesn't read like that, let's find a better wording.

```
    These functions (e.g., gr_free()), explicitly dereference the pointer
    and read the pointee.
```

So far this says that since we're reading the memory, it can't be uninitialized.  This already clashes with the documentation of /@out@/ from splint.

```
    The [[gnu::access(write_only, ...)]] attribute
    means that the pointee can be uninitialized, since it won't read it.
```

Since I didn't know where that thing is documented, I fell back to comparing to the
closest attribute that GCC offers, which is `[[gnu::access(write_only, ...)]]` (a.k.a. `__attribute__((__access__(write_only, ...)))`).  After seeing the `/@out@/` docs, it seems a 1:1 match.  I'll reword to refer to `/@out@/` and add a link to the splint docs.

```
    That's appropriate for free(3), which does not read the memory it frees
    (however, free(3) does not read _nor_ write to the memory, so it would
    be inappropriate to use write_only with it).  However, our functions
    don't behave like free(3), and they do read the memory.
```

I compared to free(3) because it's likely the reason why the author of this code was confused and thought that `/@out@/` was appropriate here.  It is appropriate in the case of free(3), and so if one doesn't think too much about it, they may assume it's also usable in these functions, since after all, they also free a resource, so they should be similar.

Maybe I can make it a bit more explicit that I mention this because I consider it to be likely the reason why `/@out@/` was (incorrectly) used in the first place in this code.

Reply:
Thanks - sounds right.